nvme.c

```diff
@@ -221,70 +217,82 @@ void *nvme_realloc(void *p, size_t len)
 	return result;
 }
 
-static void *__nvme_alloc_huge(size_t len, bool *huge)
-{
+#define HUGE_MIN 0x80000
+
+struct nvme_mem_huge {
+	size_t len;
+	bool nvme_malloc; /* p has been allocated using nvme_malloc */
 	void *p;
+};
 
-	if (!posix_memalign(&p, getpagesize(), len)) {
-		*huge = false;
-		memset(p, 0, len);
-		return p;
-	}
-	return NULL;
-}
+struct nvme_mem_huge *nvme_alloc_huge(size_t len)
+{
+	struct nvme_mem_huge *mh;
 
-#define HUGE_MIN 0x80000
+	mh = malloc(sizeof(*mh));
```

Comment:
Would be nice to avoid another heap allocation and pointer indirection. Could we just return `nvme_mem_huge` by value?
If you drop the `bool nvme_malloc` field, then it's 16 bytes and can be returned in registers.

Reply:
As pointed out below, I haven't found a way to figure how to differentiate between `mmap` allocation and the `posix_memalign` fallback when `mmap` doesn't work.

Reply:
Regardless of its size, it would be more efficient to pass the `nvme_mem_huge` on the stack rather than heap-allocating it. It would also probably simplify the code a bit as you don't need to remember to free it in a bunch of places.

Reply:
Non of the callers are in any hotpath and I would be surprised if it's really possible to measure out any performance difference between the variants. Sure in a tight loop comparing heap-allocation vs stack-allocation will result a measurable difference. But I expect that noise around loading an application and stopping in a tight loop will just hide any differences between two allocation types.

I'd rather have the focus on API design and easy/correct use of it. Thus my question here is how can you avoid the call to `nvme_free_huge`? We still need to free the allocated memory somehow or do you mean by using the cleanup hooks?

Reply:
I don't mean avoiding freeing the data memory; I mean freeing the `nvme_mem_huge` allocation itself (there are 4 `free(mh)`s in the current code).
I would suggest one of the following APIs, allowing the `nvme_mem_huge` to be placed on the stack:
```c
struct nvme_mem_huge nvme_alloc_huge(size_t len);
bool nvme_alloc_success(const struct nvme_mem_huge *);
```
or
```c
bool nvme_alloc_huge(size_t len, struct nvme_mem_huge *allocation);
```

Reply:
Yeah, it's more a convenience thing for the new API than a performance thing. Returning the descriptor by value on the stack just makes it easier to cleanup on error later. Unless you have a need for the allocation descriptor to outlive the scope that allocated it? I didn't see any use for that here, at least.

Reply:
Yeah, and the caller could always move the `nvme_mem_huge` to the heap if they needed that

Reply:
I decided to follow the existing API and return pointer directly in `nvme_alloc_huge`. Hope this makes everyone happy :)